runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMinePlugin.java

```diff
@@ -0,0 +1,136 @@
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.blastmine;
+
+import com.google.common.eventbus.Subscribe;
+import com.google.inject.Provides;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import net.runelite.api.GameObject;
+import net.runelite.api.GameState;
+import net.runelite.api.coords.WorldPoint;
+import net.runelite.api.events.ConfigChanged;
+import net.runelite.api.events.GameObjectSpawned;
+import net.runelite.api.events.GameStateChanged;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.client.ui.overlay.Overlay;
+
+import javax.inject.Inject;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+@Slf4j
+@PluginDescriptor(
+	name = "Blast Mine"
+)
+public class BlastMinePlugin extends Plugin
+{
+	@Inject
+	private BlastMineRockOverlay blastMineRockOverlay;
+
+	@Inject
+	private BlastMineOreCountOverlay blastMineOreCountOverlay;
+
+	@Getter
+	private final Set<BlastMineRock> rocks = new HashSet<>();
+
+	@Provides
+	BlastMinePluginConfig getConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(BlastMinePluginConfig.class);
+	}
+
+	@Override
+	protected void startUp() throws Exception
+	{
+		blastMineRockOverlay.updateColors();
+	}
+
+	@Override
+	public Collection<Overlay> getOverlays()
+	{
+		return Arrays.asList(blastMineRockOverlay, blastMineOreCountOverlay);
+	}
+
+	@Subscribe
+	public void onGameObjectSpawned(GameObjectSpawned event)
+	{
+		GameObject gameObject = event.getGameObject();
+
+		BlastMineRock newRock = new BlastMineRock(gameObject);
+		BlastMineRock oldRock = getRockFromCollection(gameObject);
+
+		if (newRock.getType() == null)
+		{
+			return;
+		}
+		else if (oldRock == null)
+		{
+			log.debug("Added rock, {}", newRock.getType());
+			rocks.add(newRock);
+		}
+		else if (oldRock.getType() != newRock.getType())
+		{
+			log.debug("Updated rock, {} -> {}", oldRock.getType(), newRock.getType());
+			rocks.remove(oldRock);
+			rocks.add(newRock);
+		}
+	}
+
+	@Subscribe
+	public void onGameStateChanged(GameStateChanged event)
+	{
+		if (event.getGameState() == GameState.LOADING)
+		{
+			rocks.clear();
+		}
+	}
+
+	@Subscribe
+	public void onConfigChanged(ConfigChanged event)
+	{
+		if (event.getGroup().equals("blastmineplugin"))
+		{
+			blastMineRockOverlay.updateColors();
+		}
+	}
+
+	private BlastMineRock getRockFromCollection(GameObject gameObject)
+	{
+		WorldPoint gameObjectLocation = gameObject.getWorldLocation();
+		for (BlastMineRock rock : rocks)
```

Comment:
You are using hashSet, just override equals and hashCode instead of looping like this.